sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormat.scala

```diff
@@ -152,6 +152,16 @@ trait FileFormat {
     }
   }
 
+  /**
+   * Returns whether this format supports the given [[DataType]] in read/write path.
+   *
+   * By default all data types are supported except [[CalendarIntervalType]] in write path.
+   */
+  def supportDataType(dataType: DataType, isReadPath: Boolean): Boolean = dataType match {
```

Comment:
Hm, shouldn't we better whitelist them rather then blacklist?

Reply:
Blacklist is easier.
With whitelist , we will have to validate 
``` 
BooleanType | ByteType | ShortType | IntegerType | LongType | FloatType | DoubleType |
          StringType | BinaryType | DateType | TimestampType | DecimalType
```
Of course we can have a default function to process these. But if we add a new data source which didn't support all of them, the implementation will be verbose.

Reply:
Might be easier to write but it doesn't consider if we happened to have some more types on the other hand. It should better be explicit on what we support on the other hand.

Reply:
I wrote CSV's with whitelisting before per @hvanhovell's comment long time ago. I was (am still) okay either way but might be good to leave a cc for him.

Reply:
Whitelist for all file formats is behavior change. There are external file sources like https://github.com/databricks/spark-avro , which we probably have to update the code to make it compatible.

Currently exceptions are thrown in `buildReader` / `buildReaderWithPartitionValues`/ `prepareWrite` for unsupported types. 

So overall I prefer blacklist.

Reply:
I meant the `case`s in the match in each implementation within Spark. I didn't mean about the semantic about the API itself.

Reply:
@HyukjinKwon Sorry I don't understand. Do you mean the default case is not supported?
 ```
case _  =>  false
```
But how to make all the external formats work?

Reply:
OK. I meant, leaving the default case `true`

```scala
def supportDataType(...): Boolean = dataType match {
  case _ => true
}
```

and whitelist each type within each implementation, for example, in `CSVFileFormat.scala`

```scala
def supportDataType(...) ...
    case _: StringType | ... => true
    case _ => false
```

Reply:
blacklist is easier, but whitelist is safer.